### PR TITLE
[RFC] add prettier to eslint

### PR DESCRIPTION
**Pull request by brad-decker**
### rationale
Eslint helps to make sure that everyone follows convention, but it doesn't take an opinionated approach to do so. Prettier has some stringent rules that they are rather inflexible on, but in my experience, it results in code formatting. that is universal and enforced on save. It, by default, attempts to keep code fitting within the average width of a half screen code window (80 col) but also maximizes the amount of code per line with the legibility as its primary concern.

~The following PR has been split up into commits to make the diff manageable.~  Looking for feedback on how we can modify this setup to meet the needs of the entire team.

Link to `.prettierrc.yml` in feature branch: https://github.com/MetaMask/metamask-extension/blob/prettier/.prettierrc.yml

@rekmarks pointed out in slack the changes to the eslint config, copy pasted:

Disabled/attenuated:
indent
no-mixed-operators
no-confusing-arrow
react/jsx-closing-tag-location
react/jsx-wrap-multilines
Added:
react/self-closing-comp

**NEW RULE UPDATES AFTER REBASE**
disabled: wrap-regex
attenuated: lines-around-comment (prettier has some rules around this that conflict with the default eslint rule config)

**Strategy for merging**
Once we all agree on the changes and the settings, I can split this into multiple PRs using the following strategy:

For the first PR add prettier, eslint-prettier plugin, and update rules, use overrides by folder path to disable the prettier rules, and revert the changes to the eslint rules that were required to the code. So this should result in only .eslintignore, .eslintrc, package.json and yarn.lock changing.

I can then have successive PRs for each subpath of the application so that the changes can be reviewed in chunks and if we find anything I can pause the progress of merging PRs, fix the eslint config and update any issues resulting from that in previously merged work.

The final PR would remove the overrides altogether. 

I think there would be additional steps to lift this configuration to the eslint-config package and distribute prettier with it, then we'd need to gradually adopt it in all of the other repositories. If that's a thing we want to do after using it in extension codebase.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -25,76 +25,82 @@ module.exports = {
...
-    '$': 'readonly',
+    $: 'readonly',
     document: 'readonly',
     QUnit: 'readonly',
     window: 'readonly',
   },
 
   rules: {
+    'prettier/prettier': 'error',
+    'space-before-function-paren': [
```
Comment (brad-decker) [inline .eslintrc.js:39]: prettier is opinionated about this and does not support space before, except for anonymous functions which is *forces*

File: .eslintrc.js [hunk 2]
```diff
@@ -25,76 +25,82 @@ module.exports = {
...
+      'error',
+      { anonymous: 'always', named: 'never' },
+    ],
+    indent: 'off',
```
Comment (brad-decker) [inline .eslintrc.js:43]: I tried modifying indent rules to match prettier, but the recommendation from prettier is just to disable formatting rules entirely and so I did ( and came here with this snippet to justify me taking the easy road )

File: .eslintrc.js [hunk 3]
```diff
@@ -25,76 +25,82 @@ module.exports = {
...
     'arrow-parens': 'error',
     'no-tabs': 'error',
-    'no-mixed-operators': 'error',
+    'no-mixed-operators': 'off',
```
Comment (brad-decker) [inline .eslintrc.js:47]: https://github.com/prettier/prettier/issues/187 prettier attempts to make output HUMAN readable, check out specifically this:

image

and what it now is formatted to:
image

File: .eslintrc.js [hunk 4]
```diff
@@ -25,76 +25,82 @@ module.exports = {
...
+    'no-confusing-arrow': 'off',
```
Comment (brad-decker) [inline .eslintrc.js:48]: Prettier decides what's confusing, not you, puny human

Comment (rekmarks) [inline .eslintrc.js:48]: Is there any way these could remain enabled?

File: .eslintrc.js [hunk 5]
```diff
@@ -25,76 +25,82 @@ module.exports = {
...
-    'react/jsx-curly-brace-presence': ['error', { 'props': 'never', 'children': 'never' }],
+    'react/jsx-curly-brace-presence': [
+      'error',
+      { props: 'never', children: 'never' },
+    ],
     'react/jsx-equals-spacing': 'error',
     'react/no-deprecated': 'error',
     'react/default-props-match-prop-types': 'error',
-    'react/jsx-closing-tag-location': 'error',
+    'react/jsx-closing-tag-location': [
```
Comment (brad-decker) [inline .eslintrc.js:78]: updated to match prettier's rules, but could just as easily disable

File: .eslintrc.js [hunk 6]
```diff
@@ -25,76 +25,82 @@ module.exports = {
...
+      'error',
+      { selfClosing: 'tag-aligned', nonEmpty: 'tag-aligned' },
```
Comment (brad-decker) [inline .eslintrc.js:80]: ditto.

File: .eslintrc.js [hunk 7]
```diff
@@ -25,76 +25,82 @@ module.exports = {
...
-    'react/jsx-wrap-multilines': ['error', {
-      'declaration': 'parens-new-line',
-      'assignment': 'parens-new-line',
-      'return': 'parens-new-line',
-      'arrow': 'parens-new-line',
-      'condition': 'parens-new-line',
-      'logical': 'parens-new-line',
-      'prop': 'parens-new-line',
-    }],
+    'react/jsx-max-props-per-line': [
```
Comment (brad-decker) [inline .eslintrc.js:85]: should be disabled, prettier does this.

File: .eslintrc.js (hunk 4 above)
Comment (brad-decker) [reply .eslintrc.js:48]: https://github.com/prettier/eslint-config-prettier/issues/17

Comment (brad-decker) [reply .eslintrc.js:48]: also i posted some screenies from no-mixed-operators discussion.

**Comment by brad-decker** [comment]
@whymarrh this was opened while you were out, I got some initial responses in slack about some of the changes -- would love to hear your thoughts if you get time between fires/tasks.

**Comment by whymarrh** [comment]
We should revisit this once #8982 is completed!

**Comment by whymarrh** [comment]
I've created a tracking issue for v4 of our ESLint config: https://github.com/MetaMask/eslint-config/issues/67

We should be able to put most of the config changes into that so that v4 allows us to standardize prettier across all of the repos.

**Comment by brad-decker** [comment]
@whymarrh do you want me to move over the non-controversial rule changes now in new prs to the config? Also what qualifies as non-controversial? I see you added the space before parens change already.

**Comment by whymarrh** [comment]
I don't know if makes sense to move these over now or to land them here first and move things over afterwards—I'll defer to y'all on that. Just want to leave the proverbial v4 door open for these changes.

**Comment by whymarrh** [comment]
I remember seeing a PR made to prettier itself—are we waiting on that to update this? (No rush, just tryna understand.)

**Comment by brad-decker** [comment]
@whymarrh I can champion the PR and try and get some people to review it but ultimately when or if that merges is out of our control. My personal, selfish, preference would be to accept the unfortunate change to no-mixed-operators as bookmarked by @rekmarks for now until such a time as I am able to get that PR over the finish line.

I don't have a preference for order of operations here in term of what gets merged where first, just want to make sure I'm doing everything I need to do to progress this PR the way the team feels is the most comfortable.

**Comment by brad-decker** [comment]
Also, for reference, this is the PR: https://github.com/prettier/prettier/pull/9026

**Comment by brad-decker** [comment]
🎉 rebased.

**Comment by brad-decker** [comment]
🎉 and again. I can do this all day, folks. 💻 ☕

**Comment by darkwing** [comment]
Config looks good to me!

File: .eslintrc.js (hunk 7 above)
Comment (Gudahtt) [reply .eslintrc.js:123]: This is marked as `resolved`, but is still enabled :thinking:

File: .eslintrc.js [hunk 8]
```diff
@@ -38,48 +38,100 @@ module.exports = {
...
+    'prettier/prettier': 'error',
+
+    // Our usage of spaces before *named* function parens is unusual, and
+    // doesn't match the prettier spec. prettier does not offer an option
+    // to configure this
+    'space-before-function-paren': [
+      'error',
+      { anonymous: 'always', named: 'never' },
+    ],
+    // 'lines-around-comment': 'error',
```
Comment (Gudahtt) [inline .eslintrc.js:60]: Nit: This seems to have been unintentionally left here
```suggestion
```

File: .eslintrc.js [hunk 9]
```diff
@@ -38,48 +38,100 @@ module.exports = {
...
+    // when autofixing in vscode and others.
+    'no-confusing-arrow': 'off',
+    // There is no configuration in prettier for how it stylizes regexes, which conflicts
+    // with wrap-regex.
+    'wrap-regex': 'off',
+    // Prettier handles all indentation automagically. it can be configured here
+    // https://prettier.io/docs/en/options.html#tab-width but the default matches our
+    // style.
+    indent: 'off',
+    'function-paren-newline': 'off',
```
Comment (Gudahtt) [inline .eslintrc.js:99]: Was there an explanation somewhere for why these three rules were disabled? :thinking:

File: .eslintrc.js [hunk 10]
```diff
@@ -94,66 +145,60 @@ module.exports = {
...
-      'test/**/*.spec.js',
-    ],
-    rules: {
-      // Mocha will re-assign `this` in a test context
-      '@babel/no-invalid-this': 'off',
+    {
+      files: ['test/**/*-test.js', 'test/**/*.spec.js'],
+      rules: {
+        // Mocha will re-assign `this` in a test context
+
```
Comment (rekmarks) [inline .eslintrc.js:171]: nit: This can be removed if we want to, right?
```suggestion
```

File: .prettierrc.yml [hunk 11]
```diff
@@ -0,0 +1,3 @@
+singleQuote: true
+semi: false
```
Comment (rekmarks) [inline .prettierrc.yml:2]: We should enable semicolons here or in a follow-up. Our TypeScript eslint config demands it.

Comment (brad-decker) [reply .prettierrc.yml:2]: follow up, please 😅

File: .eslintrc.js [hunk 12]
```diff
@@ -38,48 +41,108 @@ module.exports = {
...
+    ],
+    // Prettier has some opinions on mixed-operators, and there is ongoing work
+    // to make the output code clear. It is better today then it was when the first
+    // PR to add prettier. That being said, the workaround for keeping this rule enabled
+    // requires breaking parts of operations into different variables -- which I believe
+    // to be worse. https://github.com/prettier/eslint-config-prettier#no-mixed-operators
+    'no-mixed-operators': 'off',
+    // Prettier wraps single line functions with ternaries, etc in parens by default, but
+    // if the line is long enough it breaks it into a separate line and removes the parens.
+    // The second behavior conflicts with prettier. There is some guides on the repo about
```
Comment (rekmarks) [inline .eslintrc.js:88]: >  Prettier does x and y... The second behavior conflicts with prettier.

I think there's an `eslint` missing in here for the comment to make sense?

File: .prettierrc.yml (hunk 11 above)
Comment (rekmarks) [reply .prettierrc.yml:2]: Haha, okay, _fine_!

**Review by rekmarks — APPROVED** [review]
Let's get this merged before anything else happens!

**Pull request merged**